File: internal/app/auth_service.go
package app

import (
	"context"
	"errors"
	"spahtmx/internal/domain"
	"time"

	"github.com/golang-jwt/jwt/v5"
	"golang.org/x/crypto/bcrypt"
)

var (
	ErrUnauthorized = errors.New("unauthorized")
)

type AuthService struct {
	userRepo domain.UserRepository
}

func NewAuthService(userRepo domain.UserRepository) *AuthService {
	return &AuthService{
		userRepo: userRepo,
	}
}

func (s *AuthService) Login(ctx context.Context, username, password string) (domain.User, error) {
	user, err := s.userRepo.GetByUsername(ctx, username)
	if err != nil {
		if errors.Is(err, domain.ErrUserNotFound) {
			return domain.User{}, ErrUnauthorized
		}
		return domain.User{}, err
	}

	err = bcrypt.CompareHashAndPassword([]byte(user.Password), []byte(password))
	if err != nil {
		return domain.User{}, ErrUnauthorized
	}

	return user, nil
}

func (s *AuthService) GetUserByUsername(ctx context.Context, username string) (domain.User, error) {
	return s.userRepo.GetByUsername(ctx, username)
}

func (s *AuthService) GenerateToken(username, secret string) (string, error) {
	token := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.RegisteredClaims{
		Subject:   username,
		ExpiresAt: jwt.NewNumericDate(time.Now().Add(24 * time.Hour)),
		IssuedAt:  jwt.NewNumericDate(time.Now()),
	})
	return token.SignedString([]byte(secret))
}

func (s *AuthService) HashPassword(password string) (string, error) {
	bytes, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
	return string(bytes), err
}

File: src/app/auth_service.rs
use std::sync::Arc;
use crate::domain::repositories::UserRepository;
use crate::domain::models::User;
use bcrypt::{verify, hash, DEFAULT_COST};

pub struct AuthService {
    user_repo: Arc<dyn UserRepository>,
}

impl AuthService {
    pub fn new(user_repo: Arc<dyn UserRepository>) -> Self {
        Self { user_repo }
    }

    pub async fn login(&self, username: &str, password: &str) -> Result<User, Box<dyn std::error::Error + Send + Sync>> {
        let user = self.user_repo.get_by_username(username).await?;
        
        if verify(password, &user.password)? {
            Ok(user)
        } else {
            Err("Unauthorized".into())
        }
    }

    pub async fn get_user_by_username(&self, username: &str) -> Result<User, Box<dyn std::error::Error + Send + Sync>> {
        self.user_repo.get_by_username(username).await
    }

    pub fn hash_password(&self, password: &str) -> Result<String, Box<dyn std::error::Error + Send + Sync>> {
        Ok(hash(password, DEFAULT_COST)?)
    }
}
